### PR TITLE
Implement the point limit monitor

### LINKED ISSUE #6077: Number of points - Kill queries that access too many points
#6024

#6073 should be useful for implementing this limit.

### PR DESCRIPTION
**Pull request by jsternberg**
Also modified the query manager to be able to return custom error messages.

Fixes #6077.

### CODE REVIEW

**Comment by jsternberg** [comment]
@benbjohnson here's the work I was talking about for implementing the point limit iterator using the query manager instead.

File: influxql/query_manager.go [hunk 1]
```diff
@@ -162,11 +223,51 @@ func (qm *defaultQueryManager) waitForQuery(qid uint64, timeout time.Duration, c
 
 	select {
 	case <-closing:
+		qm.mu.Lock()
+		query, ok := qm.queries[qid]
+		qm.mu.Unlock()
```
Comment (benbjohnson) [inline influxql/query_manager.go]: Can you move these 3 lines to a helper function? e.g. `defaultQueryManager.Query(qid)`

File: influxql/iterator.gen.go.tmpl [hunk 2]
```diff
@@ -167,6 +164,18 @@ func (itr *{{$k.name}}MergeIterator) Close() error {
...
+		itr.heap.items = make([]*{{$k.name}}MergeHeapItem, 0, len(items))
+		for _, item := range items {
+			if item.itr.peek() == nil {
+				continue
+			}
+			itr.heap.items = append(itr.heap.items, item)
+		}
+		heap.Init(itr.heap)
+	}
+
```
Comment (benbjohnson) [inline influxql/iterator.gen.go.tmpl]: Why is this in the points monitor PR?

File: influxql/query_manager.go [hunk 3]
```diff
@@ -49,8 +56,33 @@ type QueryParams struct {
...
+	// executing the query. Optional.
+	Error *QueryError
+}
+
+// QueryError is an error thrown by the QueryManager when there is a problem
+// while executing the query.
+type QueryError struct {
+	err error
+	mu  sync.Mutex
 }
```
Comment (benbjohnson) [inline influxql/query_manager.go]: Why is the mutex on the `QueryError` and not on the `QueryTask` when you're setting the error?

File: influxql/iterator.gen.go.tmpl (hunk 2 above)
Comment (jsternberg) [inline influxql/iterator.gen.go.tmpl]: The points monitor doesn't work correctly without it. I can separate it if you want.

File: influxql/query_manager.go (hunk 3 above)
Comment (jsternberg) [inline influxql/query_manager.go]: This is something that can be done better by modifying the query executor to accept an error channel instead of a struct channel. The reason is because the mutex is surrounding the error and not the task since the way to get an error is to pass in this object to the query manager. If this object is present, it will be filled with an error.

I took this from how I've seen other libraries deal with optional metadata sections getting passed in. I know in the Consul Go client if you want metadata returned, you have to create the metadata struct and then call the function passing in that struct. It will then fill the struct you pass in rather than creating its own. This uses the same mechanism.

So in conclusion, it's because this mutex is protecting the error, not the task.

File: influxql/iterator.gen.go.tmpl (hunk 2 above)
Comment (benbjohnson) [inline influxql/iterator.gen.go.tmpl]: If it's necessary then that's fine. I just don't understand why it's added and I didn't see a comment. Why doesn't it work without it?

Comment (jsternberg) [inline influxql/iterator.gen.go.tmpl]: From the commit message where I added this commit (duplicated here in case of a rebase):

> The MergeIterator creation function would call `peek()` on the iterator
> to initialize the heap. Since this function can sometimes take a long
> time (such as a huge aggregate query on a shard), the
> `influxql.Select()` wouldn't return until the query had already been
> completed.
>
> The `influxql.Select()` call should be just the creation of the
> iterators and shouldn't calculate anything. This is important for future
> features like the point limiter that have to be initialized after the
> `influxql.Select()` call.

File: influxql/query_manager.go (hunk 1 above)
Comment (jsternberg) [inline influxql/query_manager.go]: I'll add this as a private function. I was starting to think I was duplicating that code too many times myself.

File: influxql/iterator.gen.go.tmpl (hunk 2 above)
Comment (benbjohnson) [inline influxql/iterator.gen.go.tmpl]: Can you add a comment to the code? I'm sure I won't remember that in six months. :)

Comment (jsternberg) [inline influxql/iterator.gen.go.tmpl]: Sure. I think I'll also write up a document soon with the various pitfalls I've found while working myself around the query engine.
